src/install.rs

```diff
@@ -278,6 +287,373 @@ wiping them with `wipefs -a`.\n"
+    let mut v = VerifyReader::new(&mut bf, Some(sig.as_slice()), VerifyKeys::Production)
+        .with_context(|| format!("creating verifier for {}", path))?;
+    copy(&mut v, &mut std::io::sink()).with_context(|| format!("reading {}", path))?;
+    v.verify_without_logging_failure()
+        .with_context(|| format!("verifying {}", path))?;
+    drop(v);
+    drop(bf);
+
+    f.seek(SeekFrom::Start(0))
+        .with_context(|| format!("seeking to start of {}", path))?;
+    Ok(f)
+}
+
+// XXX: copied from Zincati; need https://github.com/coreos/rpm-ostree/issues/2389
+/// JSON output from `rpm-ostree status --json`
+#[derive(Clone, Debug, Deserialize)]
+pub struct StatusJson {
+    deployments: Vec<DeploymentJson>,
+}
+
+/// Partial deployment object (only fields relevant to zincati).
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct DeploymentJson {
+    booted: bool,
+    base_checksum: Option<String>,
+    #[serde(rename = "base-commit-meta")]
+    base_metadata: BaseCommitMetaJson,
+    checksum: String,
+    // NOTE(lucab): missing field means "not staged".
+    #[serde(default)]
+    staged: bool,
+    version: String,
+    osname: String,
+}
+
+/// Metadata from base commit (only fields relevant to zincati).
+#[derive(Clone, Debug, Deserialize)]
+struct BaseCommitMetaJson {
+    #[serde(rename = "coreos-assembler.basearch")]
+    basearch: String,
+    #[serde(rename = "fedora-coreos.stream")]
+    fedora_coreos_stream: Option<String>,
+}
+
+fn get_booted_deployment() -> Result<DeploymentJson> {
+    let status_json =
+        runcmd_output!("rpm-ostree", "status", "--json").context("querying `rpm-ostree status`")?;
+    let status: StatusJson =
+        serde_json::from_str(&status_json).context("deserializing `rpm-ostree status`")?;
+    for d in status.deployments {
+        if d.booted {
+            return Ok(d);
+        }
+    }
+    bail!("no booted deployment found");
+}
+
+// XXX: generalize and add to https://github.com/coreos/stream-metadata-rust/ ? though release
+// metadata is intended to be private
+
+#[derive(Deserialize)]
+struct ReleaseMetadata {
+    architectures: HashMap<String, ReleaseArch>,
+}
+
+#[derive(Deserialize)]
+struct ReleaseArch {
+    media: ReleaseMedia,
+}
+
+#[derive(Deserialize)]
+struct ReleaseMedia {
+    metal: ReleaseMediaMetal,
+}
+
+#[derive(Deserialize)]
+struct ReleaseMediaMetal {
+    artifacts: ReleaseMediaMetalArtifacts,
+}
+
+#[derive(Deserialize)]
+struct ReleaseMediaMetalArtifacts {
+    pxe: PxeArtifacts,
+}
+
+#[derive(Deserialize)]
+struct PxeArtifacts {
+    rootfs: Artifact,
+    initramfs: Artifact,
+}
+
+#[serde_as]
+#[derive(Deserialize)]
+struct Artifact {
+    #[serde_as(as = "DisplayFromStr")]
+    location: Url,
+}
+
+fn get_fcos_live_urls(
+    stream: &str,
+    version: &str,
+    arch: &str,
+    retries: FetchRetries,
+) -> Result<(Url, Url)> {
+    let release_meta_url = format!(
```

Comment:
Isn't this covered by https://github.com/coreos/stream-metadata-rust  ?

Oh, I see the comment above.

But...hum, shouldn't we mainly support re-installing from a well known stream (i.e. usually `stable`)?

The use case of reinstalling from a specific release seems like it can be covered by having the user manually provide a URL to a build.

That would allow them to use a specific release from the FCOS prod builds (until we GC them) and would allow them to use mirrored builds more easily.

Reply:
I mentioned this when I demo'ed it but not here: the idea with using the same release was that I had a faint hope we could reuse all the artifacts we have on hand to make it a fully local experience. Obviously the rootfs squashfs would've taken some creativity. :)